### PR TITLE
ARROW-11357: [Rust]: Fix out-of-bounds reads in `take` and other undefined behavior

### PR DESCRIPTION
**Pull request by jorgecarleitao**
This PR fixes two major issues in our `take` kernel for primitive arrays.

Background

* When using `values()` from an array, it is important to remember that only certain values are not arbitrary (those whose null bit is set / no buffer).

* When reading values from an array using `array.value(i)`, it is important to remember that it currently performs no bound checks.

* `take` offers an option to deactivate bound checks, by turning an error in a panic. that option defaults to not check (i.e. `panic`)

however, `take` kernel respects none of the points above: 
* it reads and uses arbitrary indices
* it accesses out of bound values
* it does not panic when `check_bounds = false` (it instead reads out of bounds)

Specifically, it is currently doing something like the following (ignoring some details):

```rust
let indices = indices.values();
for index in indices {
    let index = index.to_usize();
    let taken_value = values.value(index)
}
```

I.e. 

* there is no check that `index` is a valid slot
* there is no check that `index < values.len()`. 

Independently, each of them is unsound. Combined, they allow for spectacular unbounded memory reads: reading from a pointer offsetted by an arbitrary value. 🤯 

This PR fixes this behavior. This PR also improves performance by 20-40% (thanks to @Dandandan and @tyrelr great suggestions).

```
Your branch is ahead of 'origin/take_fix' by 1 commit.
  (use "git push" to publish your local commits)
   Compiling arrow v4.0.0-SNAPSHOT (/Users/jorgecarleitao/projects/arrow/rust/arrow)
    Finished bench [optimized] target(s) in 1m 25s
     Running /Users/jorgecarleitao/projects/arrow/rust/target/release/deps/take_kernels-683d8fd1eeba5497
Gnuplot not found, using plotters backend
take i32 512            time:   [1.0477 us 1.0519 us 1.0564 us]                          
                        change: [-25.007% -23.339% -21.840%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  2 (2.00%) high mild
  2 (2.00%) high severe

take i32 1024           time:   [1.4949 us 1.4996 us 1.5049 us]                           
                        change: [-34.066% -31.907% -29.853%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  4 (4.00%) high mild
  3 (3.00%) high severe

take i32 nulls 512      time:   [976.65 ns 983.09 ns 991.99 ns]                                
                        change: [-45.076% -39.795% -35.157%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 9 outliers among 100 measurements (9.00%)
  1 (1.00%) low mild
  3 (3.00%) high mild
  5 (5.00%) high severe

take i32 nulls 1024     time:   [1.3249 us 1.3278 us 1.3309 us]                                 
                        change: [-32.887% -31.666% -30.524%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  3 (3.00%) high mild
  5 (5.00%) high severe
```

### CODE REVIEW

**Comment by jorgecarleitao** [comment]
On a side note, this explains why `take` was always extremely fast when I tried to replace it by `MutableArray`: `MutableArray` was safe xD

On another side note, this just re-inforces the need to mark `PrimitiveArray::value`  as `unsafe`.

File: rust/arrow/src/compute/kernels/take.rs
Comment (jorgecarleitao) [inline rust/arrow/src/compute/kernels/take.rs:1535]: this test was panicking because there was an `unwrap` on `test_take_primitive_arrays`, however, this would not allow to verify whether the panic was observed when `check_bounds: false`.

I changed this test to verify that the result is an error.

File: rust/arrow/src/compute/kernels/take.rs
Comment (jorgecarleitao) [inline rust/arrow/src/compute/kernels/take.rs:1551]: this was not passing in master, this is the one that should panic...

**Comment by jorgecarleitao** [comment]
This same behavior also exists in the boolean take :/

File: rust/arrow/src/compute/kernels/take.rs
Comment (Dandandan) [inline rust/arrow/src/compute/kernels/take.rs:423]: `> 0` might be slightly more clear here?

File: rust/arrow/src/compute/kernels/take.rs
Comment (Dandandan) [inline rust/arrow/src/compute/kernels/take.rs:374]: I am not 100% clear yet why this part itself is UB (as in reading outside bounds / uninitialized memory) as long as the null buffer is used correctly?
Or maybe you mean that it could *trigger* UB when doing OOB reads when the value itself from the array would be bigger, which would trigger `UB` later on like before? At least when calling this I think the changes definitely makes sense, I wouldn't expect a panic (or OOB!) if the value in the indices was coming from a null slot :+1:

**Comment by Dandandan** [comment]
Wouldn't it be possible to have a similar as previously implementation for `take_no_nulls` and `take_values_nulls`? I think those are the most common anyway.

**Comment by jorgecarleitao** [comment]
@Dandandan , 

> Wouldn't it be possible to have a similar as previously implementation for `take_no_nulls` and `take_values_nulls`? I think those are the most common anyway.

I am sorry, I did not understand: this PR does split the kernel in the 4 cases, and uses them accordingly depending on the null count of `values` and `indices`.

File: rust/arrow/src/compute/kernels/take.rs
Comment (jorgecarleitao) [reply rust/arrow/src/compute/kernels/take.rs:374]: I think that this becomes more academic at this point, and I do not know exactly why, but, in Rust, reading an initialized value is undefined behavior, see e.g. https://doc.rust-lang.org/std/mem/union.MaybeUninit.html

We are abusing this in a lot of places, specially when we perform SIMD operations where we disregard the null bitmap and do the op over the whole buffer. I do not have a good answer here, at this point is more like "it works".

In this particular case, though, the issue is that we are doing `*values.as_ptr().offset(index.to_usize())` where `index` comes from a null slot, which is very much an OOB read. This comment is about that: any operation on `index` in UB (even in trying to cast it to `usize`, since the previous values may represent `-1i32`, as offsets are signed).

**Comment by Dandandan** [comment]
> @Dandandan , 
> 
> > Wouldn't it be possible to have a similar as previously implementation for `take_no_nulls` and `take_values_nulls`? I think those are the most common anyway.
> 
> I am sorry, I did not understand: this PR does split the kernel in the 4 cases, and uses them accordingly depending on the null count of `values` and `indices`.

Sorry, I'll try to be more clear.

As far as I understand the undefined behavior comes from using the values from the null _indices_, so wouldn't it be possible to (mostly) recover the performance for the cases where there are either no null values at all, or only for the _values_? I think cases would occur more frequently (e.g. in DataFusion) than using the kernel with indices with nulls.

**Comment by jorgecarleitao** [comment]
Now I get your point. On the PR description, I mention to _independent_ sources of unsoundness:

* we are using indices from null slots
* we are not performing bound checks on values.

In particular, the second point above allows the following code to run with exit 0:

```rust
    #[test]
    fn test_bla() {
        let indices = UInt32Array::from(vec![1, 10, 1000, 100000]);

        let values = UInt32Array::from(vec![0, 1, 2, 3]);
        let r = take(&values, &indices, None).unwrap();
        println!("{:?}", r);
    }
```

which of course gives a meaningless result. The point is that even though this code is marked as `safe`, it is UB.

**Comment by Dandandan** [comment]
@jorgecarleitao yes, that's clear. I wouldn't expect a ~2x difference in performance in bounds checking alone, I would guess some other slowdown could be caused by the change to use the iterator in the `take_no_nulls` and  `take_values_nulls`?
Let's say we use `get_unchecked`, will it perform on par with the previous code or would it still be slower because of the other  changes?

We should (of course) fix any UB / uninitialized memory / etc. issues. I'm wondering if there would be good ways to detect it in more cases automatically, e.g. based on running all unit tests. Would this be detected by Miri, or not before as we didn't do any OOB in the tests?

**Comment by jorgecarleitao** [comment]
I think that you are right, the perf degradation comes from other parts. I will update the PR tomorrow. well spotted!

**Comment by jorgecarleitao** [comment]
@Dandandan , I pushed a fix and the benches were updated. For non-nulls, this is now `[-20,-30%]` faster, but still 2x slower when the indices have nulls, as the logic is now fundamentally different.

File: rust/arrow/src/datatypes.rs
Comment (Dandandan) [inline rust/arrow/src/datatypes.rs:381]: Do those help for performance? AFAIK inline has impact mostly across crates?  I would expect most simple functions to be inlined in Arrow itself already.

**Review by Dandandan — APPROVED** [review]
Really impressive @jorgecarleitao that with the new changes it is a performance improvement with non null indices!!!
Also the degradation in performance in take for null indices looks reasonable to me.
Maybe we could add a micro benchmark for null values + non-null indices for a primitive array as well to test changes for that case?

File: rust/arrow/src/datatypes.rs
Comment (jorgecarleitao) [reply rust/arrow/src/datatypes.rs:381]: Yep, it took me 3hs to find out why there was an 100% degradation without inlines. :(

My findings is that the compiler is being unable to optimize certain operations inside iterators when they are not `inlined`.

I even copy-pasted `memory.rs`, `bytes.rs`, `buffer.rs` and `bit_utils` to a separate repo/lib to compare the performance on the same code, and the only difference between then was `usize::try_from(u32)` vs `u32.to_usize` (via `ArrowNativeType`).

I think that we should re-visit all these functions in `ArrowNativeType`. It is not clear why they should be there.

File: rust/arrow/src/compute/kernels/take.rs
Comment (tyrelr) [inline rust/arrow/src/compute/kernels/take.rs:339]: Is it possible to use Buffer::from_trusted_len_iter here?

Comment (jorgecarleitao) [reply rust/arrow/src/compute/kernels/take.rs:339]: `values`' implements `Iterator<Item=Result<T>>`. `from_trusted_len_iter` expects `Iterator<Item=T>`, `try_from_trusted_len_iter` expects `Iterator<Item=Result<T>>`.

File: rust/arrow/src/compute/kernels/take.rs
Comment (tyrelr) [inline rust/arrow/src/compute/kernels/take.rs:337]: Undefined behavior is not supposed to be accessible from safe code...  so indices.values()[i] must never return an undefined value.

So based on that, would it be performant to do something like?
```
let values = indices.values().iter().match(|index|{
    match values.get(index) {  //I _think_ slice.get() returns an Option as None if out-of-bounds???
        Some(value) => value,
        None => if indices.is_null(index) {
            T::Native::default()
         } else {
            panic!("Out-of-bounds index {}",index)
         }
    }
```

File: rust/arrow/src/compute/kernels/take.rs
Comment (tyrelr) [reply rust/arrow/src/compute/kernels/take.rs:339]: Ah, I had missed the question mark so thought the Result wrapping was unnecessary.

File: rust/arrow/src/compute/kernels/take.rs
Comment (jorgecarleitao) [reply rust/arrow/src/compute/kernels/take.rs:337]: Good idea. I will try this.

Comment (jorgecarleitao) [reply rust/arrow/src/compute/kernels/take.rs:342]: Yeap, it fixed it. Now all benched cases are `[-30%,-20%]` faster. Thanks a lot for the idea! I updated the description.

**Comment by jorgecarleitao** [comment]
@nevi-me and @alamb , could one of you go through this? Imo it is a relatively important bug as it leads to undefined behavior.

**Comment by alamb** [comment]
@jorgecarleitao  -- I will try and review carefully later today or over the weekend.

File: rust/arrow/src/compute/kernels/take.rs
Comment (alamb) [inline rust/arrow/src/compute/kernels/take.rs:342]: I think I am missing some subtlety here --  given that `indicies` may contain nulls, shouldn't the code be checking if the current index value was null *before* looking at the value of `values.get(index)`?  Something like the following (untested):


```suggestion
    let values = indices.iter().map(|index| {
       Result::Ok(index.map(|index| values.get(index)).unwrap_or(T::Native::default()))
    });
```

I don't see how if `values.get()` returns None can be used to tell if the corresponding index was null.

**Review by alamb — APPROVED** [review]
Safer and faster code? What is not to love! Nice work @jorgecarleitao -- I went through this one carefully and it makes sense to me

I had a question about `take_indices_nulls` which I think should be addressed before merging this, but as long as it has an answer to your satisfaction, I am happy with merging this PR.

File: rust/arrow/src/compute/kernels/take.rs
Comment (jorgecarleitao) [reply rust/arrow/src/compute/kernels/take.rs:342]: Good point. I should probably document it better, as it is not trivial:

when only the indices have nulls, `take_indices_nulls`, we clone the null buffer and use it on the resulting array's validity bitmap. This is why we can ignore what happens when the index is null (apart from having to panic if a non-null index is out of bounds, to be consistent with the overall behavior of the `take` function).

Comment (alamb) [reply rust/arrow/src/compute/kernels/take.rs:342]: Ah I get it -- thank you @jorgecarleitao  that makes sense

**Comment by alamb** [comment]
@jorgecarleitao  is this one ready to merge (after rebase), do you think?

**Review by alamb — COMMENTED** [review]
I ran the tests locally after merging this branch to master, and everything looks good. Merging this one in

**Pull request closed without merging**
